File: 004/solution_004.erl
-module(solution_004).
-export([start/0]).


is_num_palindrome(Number) ->
    String = integer_to_list(Number),
    Length = length(String),
    {Left, Right} = lists:split(round(Length/2), String),
    LeftReversed = lists:reverse(Left),
    case (Length rem 2) of
        0 ->
            LeftReversed =:= Right;
        1 ->
            % Dropping the pivot character, since we only have one in an
            % odd-length list
            [_|LeftRevTail] = LeftReversed,
            LeftRevTail =:= Right
    end.


solution(From, To) ->
    Products = [X * Y || X <- lists:seq(From, To), Y <- lists:seq(From, To)],
    Palindromes = [N || N <- Products, is_num_palindrome(N)],
    LargestPalindrome = lists:last(lists:sort(Palindromes)),
    LargestPalindrome.


start() ->
    io:format("~p\n", [solution(100, 999)]).

File: 004/solution_004.py
# 2012-05-29


def is_palindrome(n):
    n = str(n)
    return n == n[::-1]


def project_euler_004(first, last):
    given_range = xrange(first, last + 1)
    products = list(set([a * b for a in given_range for b in given_range]))
    palindromes = [n for n in products if is_palindrome(n)]
    solution = max(palindromes)
    print solution


if __name__ == '__main__':
    project_euler_004(100, 999)
